src/main/java/com/yscope/logging/log4j1/AbstractBufferedRollingFileAppender.java

```diff
@@ -0,0 +1,440 @@
+   */
+  private synchronized void flushAndSyncIfNecessary () throws IOException {
+    long ts = timeSource.getCurrentTimeInMilliseconds();
+    if (ts >= flushSoftTimeoutTimestamp || ts >= flushHardTimeoutTimestamp) {
+      flush();
+      backgroundSyncThread.addSyncRequest(currentLogPath, false);
+      resetFreshnessTimeouts();
+    }
+  }
+
+  /**
+   * Periodically flushes and syncs the current log file if we've exceeded one
+   * of the freshness timeouts.
+   */
+  private class BackgroundFlushThread extends Thread {
+    @Override
+    public void run () {
+      while (true) {
+        try {
+          flushAndSyncIfNecessary();
+          sleep(timeoutCheckPeriod);
+        } catch (IOException e) {
+          logError("Failed to flush buffered appender in the background", e);
+        } catch (InterruptedException e) {
+          if (closeFileOnShutdown) {
+            logDebug("Received interrupt message for graceful shutdown of BackgroundFlushThread");
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Thread to synchronize log files in the background (by calling
+   * {@link #sync(String, boolean) sync}). The thread maintains a request queue
+   * that callers should populate.
+   */
+  private class BackgroundSyncThread extends Thread {
+    private final LinkedBlockingQueue<Request> requests = new LinkedBlockingQueue<>();
+
+    @Override
+    public void run () {
+      while (true) {
+        try {
+          Request request = requests.take();
+          if (request instanceof SyncRequest) {
+            SyncRequest syncRequest = (SyncRequest)request;
+            try {
+              sync(syncRequest.logFilePath, syncRequest.deleteFile);
+            } catch (Exception ex) {
+              logError("Failed to sync '" + syncRequest.logFilePath + "'", ex);
+            }
+          } else if (request instanceof ShutdownRequest) {
+            logDebug("Received shutdown request");
+            break;
+          }
+        } catch (InterruptedException e) {
+          // Ignore the exception since we want to continue syncing logs even
+          // in case of exceptions (the thread only shuts down when a shutdown
+          // request is made)
+        }
+      }
+    }
+
+    /**
+     * Adds a shutdown request to the request queue
+     */
+    public void addShutdownRequest () {
+      logDebug("Adding shutdown request");
+      Request shutdownRequest = new ShutdownRequest();
+      while (false == requests.offer(shutdownRequest)) {}
+    }
+
+    /**
+     * Adds a sync request to the request queue
+     * @param logFilePath Path of the log file to sync
+     * @param deleteFile Whether the log file can be deleted after syncing.
+     */
+    public void addSyncRequest (String logFilePath, boolean deleteFile) {
+      Request syncRequest = new SyncRequest(logFilePath, deleteFile);
+      while (false == requests.offer(syncRequest)) {}
+    }
+
+    private class Request {}
```

Comment:
Feel Request is a very general class name. What if the user import Request class from other places and use it in the  inherited sub class?

Reply:
This `Request` object is private class nested within another private inner class. The `Request` object will not be accessible from inherited sub class and we don't provide any way for user to directly interact with the class either. Therefore, this general name shouldn't have any conflict.